### PR TITLE
Implement advanced login

### PR DESCRIPTION
**Pull request by TheNguyen0905**

### CODE REVIEW

**Comment by TheNguyen0905** [comment]
ready

**Comment by TheNguyen0905** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -147,6 +147,20 @@ aside {
...
+  }
+}
+
+#session_remember_me {
+  width: auto;
+  margin-left: 0;
+}
+
 input, textarea, select, .uneditable-input {
   border: 1px solid #bbb;
```
Comment (huuphuc93) [inline app/assets/stylesheets/custom.scss]: use color variable

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -1,20 +1,20 @@
 class SessionsController < ApplicationController
-  def new
-  end
+  def new; end
 
   def create
     user = User.find_by email: params[:session][:email].downcase
     if user&.authenticate(params[:session][:password])
       log_in user
+      params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (huuphuc93) [inline app/controllers/sessions_controller.rb]: use " instead of '

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -3,15 +3,36 @@ def log_in user
...
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by id: user_id
+    elsif user_id = cookies.signed[:user_id]
+      user = User.find_by(id: user_id)
```
Comment (huuphuc93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/models/user.rb [hunk 4]
```diff
@@ -8,9 +9,26 @@ class User < ApplicationRecord
...
     BCrypt::Password.create(string, cost: cost)
   end
+
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
```
Comment (huuphuc93) [inline app/models/user.rb]: use update_attributes.  What difference between update_attribute and update_attributes

File: app/models/user.rb [hunk 5]
```diff
@@ -8,9 +9,26 @@ class User < ApplicationRecord
...
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
+
+  def forget
+    update_attribute(:remember_digest, nil)
```
Comment (huuphuc93) [inline app/models/user.rb]: use update_attributes

File: app/views/sessions/new.html.erb [hunk 6]
```diff
@@ -7,6 +7,10 @@
       <%= f.email_field :email, class: 'form-control' %>
```
Comment (huuphuc93) [inline app/views/sessions/new.html.erb]: use " instead of '

**Comment by huuphuc93** [comment]
commented

File: app/models/user.rb (hunk 4 above)
Comment (TheNguyen0905) [reply app/models/user.rb]: update_attributes: update all attributes if pass validates
update_attribute: update 1 attribute ignore validates

**Comment by TheNguyen0905** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 7]
```diff
@@ -2,6 +2,8 @@
 @import "bootstrap";
 
 $gray-medium-light: #eaeaea;
+$red-light: FF0000;
```
Comment (huuphuc93) [inline app/assets/stylesheets/custom.scss]: move color variables to external file like color.scss and import to this file

File: app/views/sessions/new.html.erb [hunk 8]
```diff
@@ -4,9 +4,13 @@
   <div class="col-md-6 col-md-offset-3">
     <%= form_for(:session, url: login_path) do |f| %>
```
Comment (huuphuc93) [inline app/views/sessions/new.html.erb]: remove ()

File: app/models/user.rb [hunk 9]
```diff
@@ -8,9 +9,26 @@ class User < ApplicationRecord
   has_secure_password
   validates :password, presence: true, length: {minimum: 6}
 
-  def User.digest(string)
-    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
-               BCrypt::Engine.cost
+  def User.digest string
```
Comment (huuphuc93) [inline app/models/user.rb]: class method em viết theo kiểu class << self; end nhé

**Comment by huuphuc93** [comment]
commented

**Comment by TheNguyen0905** [comment]
ready

**Comment by TheNguyen0905** [comment]
ready

File: app/assets/stylesheets/color.scss [hunk 10]
```diff
@@ -0,0 +1,3 @@
+$gray-medium-light: #eaeaea;
+$red-light: FF0000;
+$gray-light: #bbb;
```
Comment (huuphuc93) [inline app/assets/stylesheets/color.scss]: end of file

File: app/models/user.rb [hunk 11]
```diff
@@ -1,4 +1,5 @@
 class User < ApplicationRecord
+  attr_accessor :remember_token
   before_save {self.email = email.downcase}
   validates :name, presence: true, length: {maximum: 50}
```
Comment (huuphuc93) [inline app/models/user.rb]: move 50 to file Settings

File: app/models/user.rb [hunk 12]
```diff
@@ -8,9 +9,28 @@ class User < ApplicationRecord
   has_secure_password
```
Comment (huuphuc93) [inline app/models/user.rb]: move this method to after  validates :password

File: app/models/user.rb [hunk 13]
```diff
@@ -1,4 +1,5 @@
 class User < ApplicationRecord
+  attr_accessor :remember_token
   before_save {self.email = email.downcase}
   validates :name, presence: true, length: {maximum: 50}
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (huuphuc93) [inline app/models/user.rb]: move this const variable to top of this class

**Comment by huuphuc93** [comment]
commented

**Comment by TheNguyen0905** [comment]
ready

File: app/models/user.rb [hunk 14]
```diff
@@ -1,16 +1,36 @@
 class User < ApplicationRecord
-  before_save {self.email = email.downcase}
-  validates :name, presence: true, length: {maximum: 50}
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
-  validates :email, presence: true, length: {maximum: 255},
+  attr_accessor :remember_token
+  before_save {self.email = email.downcase}
+  validates :name, presence: true, length: {maximum: Settings.name_maximum}
+  validates :email, presence: true, length: {maximum: Settings.email_maximum},
             format: {with: VALID_EMAIL_REGEX},
```
Comment (huuphuc93) [inline app/models/user.rb]: indent, thụt vào 2 spaces so với  validates :email nhé

**Comment by huuphuc93** [comment]
commented
comment nhỏ

**Comment by TheNguyen0905** [comment]
ready

**Comment by huuphuc93** [comment]
merged
lgtm may be you can merged
CC: [To:2076564] Nguyen Huy Hung C (93)
[To:1818111] Nguyen Thi Huong (94)
[To:1546746] Md. Mahmudur Rahman (Mahmud)

**Pull request merged**